### PR TITLE
Create InstanceId to replace use of String

### PR DESCRIPTION
**Pull request by milleruntime**
* Create InstanceId object to replace use of String across the code
* Create new API method on InstanceOperations called getInstanceId()
* Deprecate API method InstanceOperations getInstanceID()

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/data/InstanceId.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * Get a InstanceId object for the provided canonical string. This is guaranteed to be non-null
+   *
+   * @param canonical
+   *          Instance ID string
+   * @return InstanceId object
+   */
+  public static InstanceId of(final String canonical) {
+    try {
```
Comment (DomGarguilo) [inline core/src/main/java/org/apache/accumulo/core/data/InstanceId.java:53]: Should we check that `canonical` is non null here? Similar to how `uuid` is checked in the method below?

Comment (milleruntime) [reply core/src/main/java/org/apache/accumulo/core/data/InstanceId.java:53]: This is done in the constructor of the parent class `AbstractId`

**Comment by milleruntime** [comment]
I actually think this may have already found a bug in `AuthenticationTokenIdentifier` that is being exposed by the now failing test `SaslDigestCallbackHandlerTest`.

**Review by ctubbsii — COMMENTED** [review]
I haven't done a thorough review, but I skimmed it. I like the idea.

I was thinking that we might want to make `AbstractId.toString` final, so it always returns `.canonical()`, so that it's contractually consistent, just because there's lots of places where it's logged/concatenated/etc. with other strings, and it's assuming that it's the same as `.canonical()`, but that's not necessarily guaranteed.

**Comment by milleruntime** [comment]
> I was thinking that we might want to make `AbstractId.toString` final, so it always returns `.canonical()`, so that it's contractually consistent, just because there's lots of places where it's logged/concatenated/etc. with other strings, and it's assuming that it's the same as `.canonical()`, but that's not necessarily guaranteed.

That is a good idea. It will help with the other *Id types as well.

**Review by ctubbsii — APPROVED** [review]

**Pull request merged**